Wangscape/noise/module/Bitmap.cpp

```diff
@@ -0,0 +1,116 @@
+#include "Bitmap.h"
+#include "logging/Logging.h"
+
+namespace noise
+{
+namespace module
+{
+
+namespace {
+// TODO (Serin): extract the data format from noise::RasterBase into a separate class and use it here
+static std::map<std::string, std::shared_ptr<sf::Image>> loadedImages;
+}
+
+
+Bitmap::Bitmap() : 
+    Module::Module(GetSourceModuleCount())
+{
+}
+
+const std::string & Bitmap::GetFilename() const
+{
+    return mFilename;
+}
+
+void Bitmap::SetFilename(const std::string & filename)
+{
+    // Load and process file if not already done
+    auto it = loadedImages.find(filename);
+    if (it == loadedImages.cend())
+        return;
+    else
+    {
+        auto bitmap = std::make_shared<sf::Image>();
+        // TODO get base file path from Options. How?
+        bitmap->loadFromFile(filename);
+        for (size_t y = 0; y < bitmap->getSize().y; y++)
```

Comment:
Chose `Armadillo` because it has third-order tensors. `xtensor` produces security warnings in copy operations, `DyND` doesn't compile, `CImg` has a weird license, `Eigen`'s tensor support isn't mature, `Boost.multi_array` doesn't have much support for vectorised operations.